### DISCUSSION TITLE
Rerendering Whole Network Programatically

### DISCUSSION THREAD

**Discussion by mossy426-cdc** [f5/unovis#375] [Q&A] [answered]
We're in the middle of migrating microbetrace.cdc.gov from using d3.js to Unovis (Angular).  

I've looked at documentation and gallery examples, but I don't see how I can programatically trigger rerendering of the entire network to, for instance, trigger the nodeLabel function to run on all nodes after I change another variable that the nodeLabel function considers.

For Example: 
```
 nodeLabel = (n: NodeDatum, i: number) => this.getLabelByVariable(i)
 
 getLabelByVariable( index : number) {

        // If no label variable then should be none
        if (this.widgets['node-label-variable'] == 'None') {
            return '';
        } else {
            return this.session.data.nodes[index][this.widgets['node-label-variable']]
        }
    }
```

I have another function that changes the 'this.widgets['node-label-variable']', so then I want the nodeLabel function to be called on all nodes. Right now the node/link functions will only tigger for a specific node when I click on a node on the graph. I've tried change detection to see if that would work, but it doesn't appear to do anything either.  

Any prompt help would be greatly appreaciated, and love this tool!  Thank you.

**Comment by rokotyan** [comment]
@mossy426 Excited to hear that!

As far as I remember, you can trigger a change detection in Angular to make it re-render the graph: 
- https://angular.io/api/core/ChangeDetectorRef#!#detectChanges-anchor.
- https://developapa.com/angular-rerender/

@reb-dev Can you please check how we do it?

If that doesn't help, you can try updating the entire data object like `this.nodes = [...nodes]`. This way Angular will understand that the data has been updated and needs to be rendered again. Also feel free to create an online example somewhere (e.g. CodesSandox). It'll make it easier for us to take a look and provide feedback.

**Comment by mossy426-cdc** [comment]
<img width="1186" alt="Screenshot 2024-04-10 at 10 22 20 AM" src="https://github.com/f5/unovis/assets/80353770/57ca9dcc-ac99-4002-97e4-156e55a50bf2">

Appreciate the assistance!  Yes I've tried with change detection, and tried both mark for check and detect changes after updating the entire object as you can see in the above and still not getting anything triggered.  I'll get to creating an online sample.  Thanks!

**Comment by mossy426-cdc** [comment]
https://stackblitz.com/edit/angular-zyybxk?file=src%2Fforce-graph%2Fforce-graph.component.ts

Here is an example.  I change the size variable after button click and now that the variable has been updated I want to update all the nodes.

**Comment by mossy426-cdc** [comment, marked as answer]
Figured it out! I had to reassign a new data object, rather than a nodes array.  Since the entire data object is what's being passed to the graph (not the nodes/links), you must reassign that object instead and it rerenders all nodes/links again.

That being said, the whole network rearranges itself when rerendering, which may prove frustrating when you've moved everything into a specific position.

**Reply by reb-dev** [reply]
> That being said, the whole network rearranges itself when rerendering, 

Hey @mossy426, you can use `fixNodePositionAfterSimulation` in `forceLayoutSettings` property to prevent this behavior:

```
forceLayoutSettings: {
    fixNodePositionAfterSimulation: true
}
```

**Reply by mossy426-cdc** [reply]
Thanks @reb-dev!  I am using Angular and that property doesnt exist on the GraphForceLayoutSettings interface that the forceLayoutSettings has, unless I need to update?  I went ahead and made the interface "any" so I can pass that property in, but the network still seems to rerarrange itself when changing something like the size of the nodes. 

Here is a stackblitz passing in the property:
https://stackblitz.com/edit/angular-zyybxk?file=src%2Fforce-graph%2Fforce-graph.component.ts

You can press the size button to change node size and see the network rerender.

Using Unovis v1.4 and Ang 16.

**Reply by mossy426-cdc** [reply]
@rokotyan @reb-dev Ok, so I've figured something out.  Even if you dont add the "fixNodePositionAfterSimulation" property, a node will stay in the same position **only** after you've dragged it.  

After you drag a node it will become "pinned" in a sense and when you rerender the network, that node will stay in same position but the rest of the network will rearrange.   

Think this is a good start to help see what a resolution may be.  I assume there must be a way to trigger that pin for all nodes upfront rather than just the ones you have dragged (what I'm thinking the "fixNodePositionAfterSimulation" was supposed to do?) If we can make that work in the stackblitz environment then we are golden! 

Thanks again for your time :)

**Reply by mossy426-cdc** [reply]
@rokotyan @reb-dev  Just checking in to see when/if this might be able to be looked at again?  We've halted development until we can resolve this since it's a requirement for us.  We'll most likely have to use another data viz if we can't figure this out soon since we can't progress without ensuring this is achieveable.  I'm sorry for the urgency, I understand you guys are busy!

**Reply by rokotyan** [reply]
@mossy426 We're planning to do some work on the graph soon (within a couple of weeks) and can check the problem with "fixNodePositionAfterSimulation" and Angular then. But right now as a quick workaround you can try this:
1. Graph has `onLayoutCalculated` callback config option which will return you all the nodes with their coordinates
2. One you get those, you can change `layoutType` to `GraphLayoutType.Precalculated` and feed in your data with coordinates

**Reply by rokotyan** [reply]
@mossy426 I was tweaking the Graph component today and I think I found the reason why `fixNodePositionAfterSimulation` was not working. Can you please try version ~~`1.4.1-alpha.7`~~ `1.4.1-alpha.8` and tell if it solves the problem on your end?

**Reply by rokotyan** [reply]
@mossy426 Correction: `1.4.1-alpha.8`

**Reply by mossy426-cdc** [reply]
Sweet! I will try it and let you guys know. Appreciate it

**Reply by mossy426-cdc** [reply]
```
"dependencies": {
    "@angular/animations": "16.2.12",
    "@angular/cdk": "16.0.2",
    "@angular/common": "16.2.12",
    "@angular/compiler": "16.2.12",
    "@angular/core": "16.2.12",
    "@angular/forms": "16.2.12",
    "@angular/material": "16.2.14",
    "@angular/platform-browser": "16.2.12",
    "@angular/platform-browser-dynamic": "16.2.12",
    "@unovis/angular": "1.4.1-alpha.8",
    "@unovis/ts": "1.3.5",
    "rxjs": "7.8.1",
    "tslib": "2.6.2",
    "web-worker": "^1.3.0",
    "zone.js": "0.13.3"
  },
  ```
Updated the stackblitz to 1.4.1-alpha.8 
https://stackblitz.com/edit/angular-zyybxk-v1f6cy?file=src%2Fforce-graph%2Fforce-graph.component.ts

It still says the fixNodePositionAfterSimulation isn't a property of the GraphForceLayoutSettings interface, so I made it any again just to be able to pass it in. 

Seeing the same issue.  Would I also need to update the unovis typscript from 1.3.5 to something else as well?

I will checkout the other workaround as well

**Reply by rokotyan** [reply]
@mossy426 I think I know what the problem might be. The versions of  `@unovis/angular` and `@unovis/ts` should match, and you have `"@unovis/angular": "1.4.1-alpha.8"` and `"@unovis/ts": "1.3.5"`

**Reply by rokotyan** [reply]
Basically `@unovis/angular` is just an adaptor, the main codebase is `@unovis/ts`

**Reply by mossy426-cdc** [reply]
Oh, That was it!  Thanks so much for your help guys! :)

**Reply by rokotyan** [reply]
@mossy426 Nice, you're welcome

**Comment by rokotyan** [comment]
@mossy426 Right! Somehow I didn't think about it. Glad you were able to solve the problem.

**Comment by rokotyan** [comment]
@reb-dev Can you convert this into a discussion please so the others can see it.
